### PR TITLE
dev/core#1494 & dev/translation#65 call brickmoney number formatting, where supported

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fix php 7.4 support issue on money formatting - switch to function compatible with 7.4

Before
---------------------------------------
Functions that call money formatting all eventually wind up in the function formatNumericByFormat which calls a php function deprecated in php 7.4

After
----------------------------------------
Functions that call money formatting all eventually wind up in the function formatUSLocaleNumericRounded which uses brickmoney unless unavailable

Technical Details
----------------------------------------
The main blocker to switching to brick money was actually the valueFormat parameter
- now that it cannot be passed in we can switch to using the function that will
prefer the brickmoney function (but fall back to the old one).

A tangential issue is using the handling that brick money offers us better and cleaning up our code.
We have gotten a bit out of sorts on this second part - because we haven't found a path to deprecate our
currency separates yet and got a bit mixed up TBH.

This PR renames a function that is always specifiying US as the locale to make that clear. As a protected function that should be safe and it should make space for the right function to be added filling that space.

The challenge we have is that if we let brickmoney handle locale we need to not do our own currency separator handling. The issue we hit recently was that created a situation where both could be handled. We can either

1) :rocket: cold switch - start using the site locale instead of the configured separators
2) :confused: soft switch - add a new setting 'money locale' and if present then replaceCurrencySeparators would do an early return, making no change. Sites without the setting would get the old formatting. This could be transitional.
3) :-1: don't switch - love the bomb and decide currencty separator settings got us this far

Comments
----------------------------------------
I'm gonna come up with some emojis for an emoji poll

### CODE REVIEW

File: CRM/Utils/Money.php [hunk 1]
```diff
@@ -86,7 +86,7 @@ public static function format($amount, $currency = NULL, $format = NULL, $onlyNu
       CRM_Core_Error::deprecatedWarning('Passing empty currency to CRM_Utils_Money::format is deprecated if you need it for display without currency call CRM_Utils_Money::formatLocaleNumericRounded');
     }
 
-    $amount = self::formatNumericByFormat($amount);
+    $amount = self::formatUSLocaleNumericRounded($amount, 2);
```
Comment (eileenmcnaughton) [inline CRM/Utils/Money.php:89]: This is the actual change - the rest just renames a protected function to what it currently does

**Comment by seamuslee001** [comment]
@mattwire @monishdeb @JoeMurray @KarinG @mlutfy @agileware-justin I think you all will have opinions on this

**Comment by eileenmcnaughton** [comment]
Note I also commented on the most recent related issue to try to get eyes/testers

**Comment by eileenmcnaughton** [comment]
ALso note the emoji poll / setting discussion is about next steps rather than about the change in this PR

**Comment by eileenmcnaughton** [comment]
We got test fails as this represented a change in rounding. I switched the rounding mode - there are a few but I found this  bit compelling "Note that this is the rounding mode  that statistically minimizes cumulative error when applied repeatedly over a sequence of calculations."


/**
     * Rounds towards the "nearest neighbor" unless both neighbors are equidistant, in which case rounds towards the even neighbor.
     *
     * Behaves as for HALF_UP if the digit to the left of the discarded fraction is odd;
     * behaves as for HALF_DOWN if it's even.
     *
     * Note that this is the rounding mode that statistically minimizes
     * cumulative error when applied repeatedly over a sequence of calculations.
     * It is sometimes known as "Banker's rounding", and is chiefly used in the USA.
     */
    public const HALF_EVEN = 9;

**Comment by eileenmcnaughton** [comment]
There was a test fail on formatting '-.1'  which updating brick money fixed

**Comment by KarinG** [comment]
I don’t think Banker’s rounding is what we want.

Numbers like 10.5 will round down to 10, but 11.5 will round up to 12 -> massively confusing. 

I vote for normal rounding / standard math.

**Comment by mlutfy** [comment]
If not too much work, I like the idea of a new setting for soft-launch, as it would be LeXiM-ish, and allow us to test with less-critical production sites.

**Comment by eileenmcnaughton** [comment]
@KarinG I thought there was such a thing as 'standard rounding' unitl I discovered there were about 8 to choose from!

**Comment by KarinG** [comment]
@eileenmcnaughton - that’s bananas - there is only one way to round to get full marks.

**Comment by demeritcowboy** [comment]
Am I understanding this right: Before the patch, it was rounding in a mysterious locale-specifc way which in unit tests was en_US with bankers rounding. Now it standardizes on 2 digits in all locales and to keep it the same in unit tests it sets it to bankers rounding? I'm trying to understand how many variables there are here.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy it was using CEILING but that was not right so now we have to choose one of the other 8

https://github.com/brick/math/blob/master/src/RoundingMode.php

**Comment by demeritcowboy** [comment]
Ok but I think there's still another thing happening here though. Before the patch if $onlyNumber is false it didn't do rounding. Now it always does rounding. So there's two questions:

1. Should it always do rounding regardless of params and the non-rounding before was wrong, or should it never do rounding, or should the rounding depend on params like before.
2. If it does do rounding in at least one situation, what method should it use. HALF_UP is the one I would be expecting.

Fancy table follows.

| When | Thing | Result |
| ---- | ---- | ---- |
| BEFORE | cv ev "echo CRM_Utils_Money::format('12.305');" | $ 12.305 |
| AFTER | cv ev "echo CRM_Utils_Money::format('12.305');" | $ 12.30 |
| BEFORE | cv ev "echo CRM_Utils_Money::format('12.315');" | $ 12.315 |
| AFTER | cv ev "echo CRM_Utils_Money::format('12.315');" | $ 12.32 |
| BEFORE | cv ev "echo CRM_Utils_Money::format('12.305', null,null,true);" | 12.31 |
| AFTER | cv ev "echo CRM_Utils_Money::format('12.305', null,null,true);" | 12.30 |
| BEFORE | cv ev "echo CRM_Utils_Money::format('12.315', null,null,true);" | 12.32 |
| AFTER | cv ev "echo CRM_Utils_Money::format('12.315', null,null,true);" | 12.32 |

**Comment by eileenmcnaughton** [comment]
@demeritcowboy ok - $only Number is in the process of being deprecated - so for consistency I guess we want no rounding - we can achieve that by passing in the length of the value I guess

**Comment by eileenmcnaughton** [comment]
I've updated it so that this PR should no longer make the changes in rounding behaviour that @demeritcowboy identified ie

1) if onlyNumber is false no rounding takes place
2) if onlyNumber is true it is rounded as before (using a non-preferred but unchanged rounding method).

The discussion about rounding method can still happen on this PR but it is non-blocking to merging

**Comment by jaapjansma** [comment]
I just did a brief glimpse of this PR and what freightens me is that the function `formatUSLocaleNumericRounded` is used and I am not sure what it does but it looks like we are hard coding us locale settings into the code. 
Also the function  `getDecimalPlacesForAmount` uses the '.' as a decimal separator. In the netherlands we use € 1.234,56 not sure how that is handled.

**Comment by ErikHommel** [comment]
I would have the same concern as @jaapjansma. I do not understand all the details but hard coding US settings seems like a step I do not like?

**Comment by eileenmcnaughton** [comment]
@ErikHommel @jaapjansma it's what we are already doing - the format function currently called casts to the US locale - I renamed the function to make that explicit. Then we swap out the currency separators which, as I noted above - we want to stop doing

Note that the money amount coming into this function is in the US format - so the decimal separator is '.' here - however, please test it to confirm it works for you

File: CRM/Utils/Money.php [hunk 2]
```diff
@@ -85,8 +85,8 @@ public static function format($amount, $currency = NULL, $format = NULL, $onlyNu
     if ($currency === ' ') {
       CRM_Core_Error::deprecatedWarning('Passing empty currency to CRM_Utils_Money::format is deprecated if you need it for display without currency call CRM_Utils_Money::formatLocaleNumericRounded');
     }
-
-    $amount = self::formatNumericByFormat($amount);
+    $precision = self::getDecimalPlacesForAmount((string) $amount);
+    $amount = self::formatUSLocaleNumericRounded($amount, $precision > 2 ? $precision : 2);
```
Comment (mattwire) [inline CRM/Utils/Money.php:89]: Why are we hardcoding a minimum of 2 decimal places here? There are quite a few currencies that have 0 decimal places - see eg. https://developer.cybersource.com/library/documentation/sbc/quickref/currencies.pdf

Comment (mattwire) [reply CRM/Utils/Money.php:89]: Setting 0 instead of 2 would be ok here?

Comment (eileenmcnaughton) [reply CRM/Utils/Money.php:89]: Yeah - oddly enough there are still tests failing with this change - suggesting that we were rounding to 2 decimal places more often than it seemed - despite @demeritcowboy's test showing us not doing so ....

Comment (eileenmcnaughton) [reply CRM/Utils/Money.php:89]: It's actually the line

```
     $amount = money_format( '%!i', $amount);
```

that is doing the rounding

Comment (demeritcowboy) [reply CRM/Utils/Money.php:89]: Oh, of course, there is no money_format on windows.

Comment (eileenmcnaughton) [reply CRM/Utils/Money.php:89]: @demeritcowboy hmm - so rounding to 2 places would be consistent with the intent of the code I believe

Comment (demeritcowboy) [reply CRM/Utils/Money.php:89]: Yes it seems. On one unix machine I'm looking at it appears to be 2 places using, oddly, HALF_DOWN.

**Comment by mattwire** [comment]
This one seems particularly broken:
```
CRM_Contribute_Form_ContributionTest::testReSubmitSaleTaxAlteredAmount with data set #0 ('.')
Failed asserting that '220000000.00' matches expected 22000.
```

**Comment by eileenmcnaughton** [comment]
@mattwire yeah that might point to pre-exising weirdness being flushed out - I'm gonna switch back to rounding to 2 as it seems that is the expectation of the existing code.

**Comment by eileenmcnaughton** [comment]
OK pushed up with rounding to 2 - I switching to 'HALF::UP' rounding - which is 
"the rounding mode commonly taught at school"

as opposed to 

"the rounding mode that statistically minimizes cumulative error when applied repeatedly over a sequence of calculation"

since the only person to express an opinion preferred intuitiveness over cumulative accuracy.

We could revisit this specifically with regards to amounts used in financial calculations later on if need be.

**Comment by demeritcowboy** [comment]
Well, this does seem to fix the issue where contribution pages on windows have always looked like this (and apparently on freebsd):

Untitled

**Comment by eileenmcnaughton** [comment]
Pasting in the relevant comments from the dev digest

The major blocker to full support for php 7.4 has been that we use a deprecated currency library. This does not cause a problem for sites with production-appropriate php error settings but it is a blocker for us having unit-test coverage which is pretty much how we define ‘supported’. In order to switch to our preferred replacement library we had to deprecate and finally remove the money format setting which is not supported by brickmoney. Having done that, we now have a PR open that switches us over and should result in no change 

We did hit some issues in 5.34 (or earlier but only noticed in 5.34) around money formatting. Our old approach was ‘windows-inclusive’. Since the money_format library wasn’t on Windows we only partially used it - really it was only doing rounding as it turns out and we were only ever telling it to format in US style. The replacing of currency separators was being done by our own magic, as was the addition of the currency symbol. When we started calling brickmoney we were letting it do locale-based formatting and then doing our own on top. To fix this in 5.35 we told brickmoney ‘just format everything in US style like we used to’ and we’ll swap it ut.

Our current approach means we don’t get the full benefit of brickmoney - which understands that sometimes the currency symbol goes first, and sometimes it goes last and it determines the currency separators based on knowing the country and locale. Basically it’s smarter and better - but it’s clear there are migration challenges. There is an emoji poll on this question  but the leading (1-nil) proposal is to add a new setting (money_locale) and if this is set then the settings ‘thousand_separator’, ‘decimal_separator’ and ‘moneyFormat’ would be ignored.

A last point of interest on brick money is that if offers multiple types of rounding - one of which is ‘bankers rounding’ which ‘statistically minimizes cumulative error when applied repeatedly over a sequence of calculations’ - something that might come in handy (discussion on the same PR)

**Comment by eileenmcnaughton** [comment]
@mattwire @seamuslee001 this was in the dev digest a week ago & was previously tested by @demeritcowboy & discusssions are addressed.

I think we can merge this now & maybe see where we are at in terms of php7.4 unit tests with it merged

**Comment by seamuslee001** [comment]
I'm going to merge this as it seems it is no worse than before

**Pull request merged**